### DISCUSSION TITLE
Enabling foreign_keys produces SqliteException 787

### DISCUSSION THREAD

**Discussion by patrikbadanjak** [simolus3/drift#2730] [Q&A]
Turning on the foreign_keys pragma inside onUpgrade or beforeOpen methods produces SqliteException 787. What causes this and how do I mitigate it? The database is running on the separate isolate as per the documentation.

**Comment by simolus3** [comment]
> What causes this

This error code indicates that a [foreign key constraint](https://www.sqlite.org/rescode.html#constraint_foreignkey) has failed. The cause is likely that you already have a `REFERENCEs` clause in a table and there are rows for which the reference is invalid. You can use the [foreign key check](https://www.sqlite.org/pragma.html#pragma_foreign_key_check) pragma for that before enabling foreign keys:

```dart
final rows = await customSelect('pragma foreign_key_check;').get();
for (final row in rows) {
  print('Foreign key mismatch: ${row.data}');
}
```

> how do I mitigate it

You will have to find the offending rows and fix them. How to do that depends on your application, it might make sense to delete them or maybe they need to be updated or the row they're referencing is missing and needs to be inserted.
